language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs

```diff
@@ -104,8 +103,8 @@ bindTuple ann names (Tuple elTys term) body = liftQuote $ do
 -- | Given an arity @n@, create the n-ary product type.
 --
 -- @\(T_1 :: *) .. (T_n :: *) . all (R :: *) . (T_1 -> .. -> T_n -> R) -> R@
-getBuiltinProdN :: MonadQuote m => Int -> m (Type TyName ())
-getBuiltinProdN arity = do
+prodN :: Int -> Type TyName ()
+prodN arity = runQuote $ do
```

Comment:
We can create a CAF that is an infinite list that contains products of all arities and then simply take the `i`th element of a list, where `i` is an arity. This way we avoid regenerating products and can share their definitions.